posthog/models/alert.py

```diff
@@ -128,44 +165,67 @@ def evaluate_condition(self, calculated_value) -> list[str]:
-    ) -> tuple["AlertCheck", list[str]]:
-        """Add a new AlertCheck, managing state transitions and cooldown."""
-        matches = self.evaluate_condition(calculated_value) if calculated_value is not None else []
-        targets_notified = {}
-
-        # Determine the appropriate state for this check
-        if matches:
-            if self.state != "firing":
-                # Transition to firing state and send a notification
-                check_state = "firing"
-                self.last_notified_at = datetime.now(UTC)
-                targets_notified = {"users": list(self.subscribed_users.all().values_list("email", flat=True))}
-            else:
-                check_state = "firing"  # Already firing, no new notification
-                matches = []  # Don't send duplicate notifications
+        self, *, aggregated_value: Optional[float], error: Optional[dict] = None
+    ) -> tuple["AlertCheck", list[str], Optional[dict], bool]:
+        """
+        Add a new AlertCheck, managing state transitions and cool down.
+
+        Args:
+            aggregated_value: result of insight calculation compressed to one number to compare against threshold
+            error: any error raised while calculating insight value, if present then set state as errored
+        """
+
+        targets_notified: dict[str, list[str]] = {}
+        breaches = []
+        notify = False
+
+        if not error:
+            try:
+                breaches = self.evaluate_condition(aggregated_value) if aggregated_value is not None else []
+            except Exception as err:
+                # error checking the condition
+                error = {
+                    "message": f"Error checking alert condition {str(err)}",
+                }
+
+        if error:
+            # If the alert is not already errored, notify user
+            if self.state != AlertState.ERRORED:
+                self.state = AlertState.ERRORED
+                notify = True
+        elif breaches:
+            # If the alert is not already firing, notify user
+            if self.state != AlertState.FIRING:
+                self.state = AlertState.FIRING
+                notify = True
         else:
-            check_state = "not_met"
-            self.state = "inactive"  # Set the Alert to inactive if the threshold is no longer met
-            # Optionally send a resolved notification
+            self.state = AlertState.NOT_FIRING  # Set the Alert to not firing if the threshold is no longer met
+            # TODO: Optionally send a resolved notification when alert goes from firing to not_firing?
+
+        now = datetime.now(UTC)
+        self.last_checked_at = datetime.now(UTC)
+
+        # IMPORTANT: update next_check_at according to interval
+        # ensure we don't recheck alert until the next interval is due
+        self.next_check_at = (self.next_check_at or now) + alert_calculation_interval_to_relativedelta(
```

Comment:
Yes, we use a transaction below, so the `next_check_at` is only updated if we are able to check and send notifications without errors.
```
@transaction.atomic
def check_alert(alert_id: str) -> None:
```

`What stops an alert that takes a long time to be calculated from being queued to compute multiple times ` -> Great spot, had missed this. Especially now that we compute hourly alerts on a rolling basis. Have added a check so that before computing, we see if alert still needs to be computed

```
    now = datetime.now(UTC)
    if alert.next_check_at > now:
        logger.warning(
            """Alert took too long to compute or was queued too long during which it already got computed. 
            So not attempting to compute it again until it's due next""",
            alert=alert,
        )
        return
```

Reply:
Maybe i'm confused - what is this checking for?

Imagine that an insight takes 10 minutes to get computed.

We poll for alerts every 2 minutes - what stops the insight from getting calculated 5 times?

Reply:
You're right, I was just guarding against queuing the same task multiple times as didn't think trend insights would take long 🤦🏽 but have added back a flag for `is_calculating` which is set/unset when alert check calculation starts. This will prevent recalculation in this scenario